### PR TITLE
 Migrate MostViewedMacro from livetable to livedata #21 

### PR DESCRIPTION
**Pull request by Farcasut**
[Bug Fixes:]
-Fixed LiveData pagination.
-Improved LiveData filtering. Added a method to filter text based on user input. For now, title filtering searches the URL, not the page title.
-Fixed LiveData sorting issues by removing $request.reqNo != '1'.

New Features:
-Changed title property to show both page title and link to the page.
-More tests
Preview:
image

### CODE REVIEW

**Comment by Farcasut** [comment]
https://github.com/xwikisas/application-analytics/assets/92780090/1fc0230d-692f-4d6c-b720-b13c3254fcb5

File: application-analytics-default/src/test/java/com/xwiki/analytics/internal/MostViewedJsonNormaliserTest.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public void testObjectJSONResponse() throws Exception
         setupURLs("http://localhost:8080/xwiki/bin/view/Analytics/",
             "http://localhost:8080/xwiki/bin/view/Analytics/Code/Macros/MostViewedPage");
         assertEquals(node.get("ResponseObjectJSON"),
-            mostViewedJsonNormaliser.normaliseData(node.get("ObjectJSON").toString()));
+            mostViewedJsonNormaliser.normaliseData(node.get("ObjectJSON").toString(), null));
```
Comment (oanalavinia) [inline application-analytics-default/src/test/java/com/xwiki/analytics/internal/MostViewedJsonNormaliserTest.java:69]: You should add tests for filters as well

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-analytics-default/src/test/java/com/xwiki/analytics/internal/MostViewedJsonNormaliserTest.java [hunk 2]
```diff
@@ -62,21 +63,33 @@ public class MostViewedJsonNormaliserTest
     @Test
     public void testObjectJSONResponse() throws Exception
```
Comment (oanalavinia) [inline application-analytics-default/src/test/java/com/xwiki/analytics/internal/MostViewedJsonNormaliserTest.java:64]: You don't have to specify the `test` part, is redundant since you have the annotation.
What it is done is to name a test after the method you are testing, like `methodName` + `particularities`. Here maybe you could have `normalizeDataWithJSONResponse`. The same name should be used for the JSON file used, to do the matching easier when something is checked.

The same applies to the tests below.

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java [hunk 3]
```diff
@@ -61,105 +57,42 @@
...
-        // field called 'date'. This 'date' field will be set to 'N/A' when Matomo returns an array instead of an
-        // object. For both type of formats, the label field is also altered in order to contain the full page name
-        if (jsonRoot.isArray()) {
-            processArrayNode(jsonRoot);
-        } else {
-            jsonRoot = processObjectNode(jsonRoot);
-        }
-        return jsonRoot;
-    }
+        if (matchesAllFilters(objNode, filters)) {
```
[the remark this answers is not in the document]
Comment (Farcasut) [reply application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java:85]: I think it is a good idea to make the processNode handle only the process part and the filters be verified outside of it. This way the code might become clearer and easier to understand in the future.

File: application-analytics-default/src/test/java/com/xwiki/analytics/internal/MostViewedJsonNormaliserTest.java [hunk 4]
```diff
@@ -57,44 +66,111 @@ public class MostViewedJsonNormaliserTest
...
+    {
+        readJSONS("/mostViewedPages/jsonWithoutURL.json");
+        assertEquals(node.get("ArrayWithoutURLResponse"),
+            mostViewedJsonNormaliser.normaliseData(node.get("ArrayWithoutURL").toString(), null));
+    }
+
+    private void setupAnyURL() throws Exception
+    {
+        ResourceType resourceType1 = mock(ResourceType.class);
+        when(resourceTypeResolver.resolve(any(ExtendedURL.class), eq(Collections.emptyMap()))).thenReturn(null);
```
Comment (oanalavinia) [inline application-analytics-default/src/test/java/com/xwiki/analytics/internal/MostViewedJsonNormaliserTest.java:167]: Why not have this in a [@Before](https://github.com/xwikisas/application-licensing/blob/d168fb88fc0d121bf95e769ea21c55c00bebe5a6/application-licensing-licensor/application-licensing-licensor-api/src/test/java/com/xwiki/licensing/internal/TrialLicenseGeneratorTest.java#L85) method?

File: application-analytics-default/src/test/java/com/xwiki/analytics/internal/configuration/DefaultAnalyticsConfigurationTest.java [hunk 5]
```diff
@@ -45,24 +49,108 @@ public class DefaultAnalyticsConfigurationTest
...
         assertEquals("13.192.200.19", this.defaultAnalyticsConfiguration.getRequestAddress());
     }
+
+    /**
+     * Will test to see if the function will throw an error when the request address is null
+     */
+    @Test
+    public void missingRequestAddress()
+    {
+        ReflectionUtils.setFieldValue(this.defaultAnalyticsConfiguration, "logger", this.logger);
```
Comment (oanalavinia) [inline application-analytics-default/src/test/java/com/xwiki/analytics/internal/configuration/DefaultAnalyticsConfigurationTest.java:82]: You should add this in a @Before method

File: application-analytics-default/src/test/java/com/xwiki/analytics/internal/configuration/DefaultAnalyticsConfigurationTest.java [hunk 6]
```diff
@@ -45,24 +49,108 @@ public class DefaultAnalyticsConfigurationTest
...
+            this.defaultAnalyticsConfiguration.getAuthenticationToken();
+        });
+        assertEquals("The authToken is missing.", exception.getMessage());
+    }
+
+    /**
+     * Will test to see if the function will throw an error when the request address is empty
+     */
+    @Test
+    public void returnDefaultValueRequestAddress()
```
Comment (oanalavinia) [inline application-analytics-default/src/test/java/com/xwiki/analytics/internal/configuration/DefaultAnalyticsConfigurationTest.java:119]: Maybe `getAddressWithDefaultValue`? Below as well

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 7]
```diff
@@ -61,5 +61,19 @@
...
+&lt;a href=$link&gt;$text&lt;/a&gt;
+#end
+
+#macro (subSet $object $start $end $resultList)
+  #set($resultList = [])
+  #set ($resultList = [])
+  #set ($end = $mathtool.sub($end,1))
+  #foreach($i in [$start..$end])
+    #set($element = $object.get($i))
+    #set($dummy = $resultList.add($element))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:75]: Usually `discard` is used as the dummy variable

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml [hunk 8]
```diff
@@ -78,18 +87,25 @@
...
     "offset": $mathtool.add($offset, 1),
     "reqNo": $numbertool.toNumber($request.reqNo).intValue(),
     "rows": []
   })
-  #foreach ($currentEntry in $analyticsResult)
+  #foreach ($currentEntry in $resultList)
     #set ($url = $currentEntry.get('url').asText())
     #set ($discard = $results.rows.add({
-      'title' : $currentEntry.get('label').asText(),
+      'test': $resultList,
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml:107]: Is this used?

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/AbstractJsonNormaliser.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        // object. For both type of formats, the label field is also altered in order to contain the full page name
+        if (jsonRoot.isArray()) {
+            return processArrayNode(jsonRoot, filters);
+        } else {
+            return processObjectNode(jsonRoot, filters);
+        }
+    }
+
+    /**
+     * This method processes each entry to append an empty date to it.
```
Comment (oanalavinia) [inline application-analytics-default/src/main/java/com/xwiki/analytics/internal/AbstractJsonNormaliser.java:79]: You should rephrase it to not have the `this method` part and maybe add more info (why the empty date?)

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/AbstractJsonNormaliser.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    /**
+     * Handle the scenario where Matomo returns an object with dates as keys and arrays of JSON objects as values. This
+     * function extracts the date from the key and adds it to each page. Ultimately, it returns an array of JSON
+     * objects.
+     *
+     * @param jsonNode
+     * @param filters holds the criteria for filtering a dataset
+     * @return array of jsons
```
Comment (oanalavinia) [inline application-analytics-default/src/main/java/com/xwiki/analytics/internal/AbstractJsonNormaliser.java:103]: `jsonNode` and `return` could have more details

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/AbstractJsonNormaliser.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    /**
+     * Process the current node and add it to the final array of jsons.
+     *
+     * @param currentNode the current json that has to be processed
+     */
+    protected JsonNode processNode(JsonNode currentNode)
+    {
+        return currentNode;
```
Comment (oanalavinia) [inline application-analytics-default/src/main/java/com/xwiki/analytics/internal/AbstractJsonNormaliser.java:143]: Actually, I think you can have it as an abstract method

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java [hunk 12]
```diff
@@ -61,103 +55,34 @@
...
-        }
-        return jsonRoot;
-    }
-
     /**
-     * This method processes each entry to append an empty date to it.
+     * Process the current node and add it to the final array of jsons.
      *
-     * @param jsonNode an array of jsons
+     * @param currentNode the current json that has to be processed
```
Comment (oanalavinia) [inline application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java:76]: `JSON`

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java [hunk 13]
```diff
@@ -61,103 +55,34 @@
...
-            processArrayNode(jsonRoot);
-        } else {
-            jsonRoot = processObjectNode(jsonRoot);
-        }
-        return jsonRoot;
-    }
-
     /**
-     * This method processes each entry to append an empty date to it.
+     * Process the current node and add it to the final array of jsons.
```
Comment (oanalavinia) [inline application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java:74]: You are not adding it here to the final list of JSONs. I think you could add some details about the exact processing you are doing (altering the label)

File: application-analytics-default/src/main/java/com/xwiki/analytics/internal/MatomoAnalyticsManager.java [hunk 14]
```diff
@@ -72,20 +74,44 @@ public class MatomoAnalyticsManager implements AnalyticsManager
...
+
+    /**
+     * Validates that the request parameters are not null.
+     *
+     * @param parameters a list of key, value pairs
+     * @throws RuntimeException if parameters are null
+     */
+    private void validateParameters(Map<String, String> parameters)
+    {
+        if (parameters == null) {
```
Comment (oanalavinia) [inline application-analytics-default/src/main/java/com/xwiki/analytics/internal/MatomoAnalyticsManager.java:96]: You should just do this check above, no need of a separate method

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml [hunk 15]
```diff
@@ -62,10 +62,19 @@
...
-  #if ($request.sort &amp;&amp; $request.reqNo!='1')
-    #set($sort = $keys.get($request.sort))
-    #set($order = $request.dir)
+  #if ($request.sort)
+    #set ($sort = $keys.get($request.sort))
+    #set ($order = $request.dir)
+  #end
+
+  #set ($filters={})
+  #foreach ($key in $keys.keySet())
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml:71]: I'll add a comment for the `keys` usage. Maybe also use a more specific name `matomoKeysMapping`, or just `keysMapping`?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 16]
```diff
@@ -61,5 +61,19 @@
...
   #set ($date = $startDate + ',' + $endDate)
 #end
+
+#macro (createLinkElement $label $link)
+&lt;a href = $link&gt;$label&lt;/a&gt;
+#end
+
+#macro (subSet $object $start $end $resultList)
+  #set ($resultList = [])
+  #set ($resultList = [])
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:71]: Duplicate

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml [hunk 17]
```diff
@@ -62,10 +62,19 @@
...
+    #set ($value = $request.get($key))
+    #if ($value &amp;&amp; $value != '')
+      $filters.put($keys.get($key), $request.get($key))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml:74]: Missing `#set($discard =..`

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml [hunk 18]
```diff
@@ -62,10 +62,19 @@
...
   #set ($order = 'desc')
-  #if ($request.sort &amp;&amp; $request.reqNo!='1')
-    #set($sort = $keys.get($request.sort))
-    #set($order = $request.dir)
+  #if ($request.sort)
+    #set ($sort = $keys.get($request.sort))
+    #set ($order = $request.dir)
+  #end
+
+  #set ($filters={})
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPagesJSON.xml:70]: Missing spaces

**Review by oanalavinia — CHANGES_REQUESTED** [review]
Almost done:)

**Review by oanalavinia — APPROVED** [review]
